### ISSUE TITLE
Feature Request: Respect user's typed case when offering suggestions

### ISSUE THREAD

**Issue by Joshfindit** [robbyrussell/oh-my-zsh#4247]
Example: the following exist:

```
./Temp1
./temp2
```

Then type: ```cd te<tab>```. zsh fills in ```temp``` then stops. Press ```<tab>``` again, and the suggestions come up as ```Temp1/ temp2/```

Coming from bash, I would expect that specifying the case of the first couple of letters would be enough to identify the desired autocomplete. 

In oh-my-zsh, is it possible to have
```cd te<tab><tab>``` autocomplete ```temp2```, then offer ```temp2/ Temp1/``` on an additional tab press?
Similarly;
have ```cd Te<tab><tab>``` autocomplete ```Temp1```, then offer ```Temp1/ temp2/```

**Comment by mcornella** [comment]
You need to add `setopt menu_complete` to your zshrc file. We disable that option in [`lib/completion.zsh`](https://github.com/robbyrussell/oh-my-zsh/blob/192de6bcffb0294e19f4203f6f7dc1a7f3e427be/lib/completion.zsh#L3).

**Comment by Joshfindit** [comment]
```setopt menu_complete``` doesn't address the different cases in the way I described.

What it does do is:
Type: ```cd te<tab>```. zsh fills in ```Temp1/```. Press ```<tab>``` again, and the suggestions come up as ```Temp1/ temp2/```

Instead, if user types: ```cd te<tab>```. zsh fills in ```temp2/```. Press ```<tab>``` again, and the suggestions come up as ```temp2/ Temp1/ ```

Another use-case: if ```./temp2``` and ```./temp3``` existed, then with ```cd te<tab>``` zsh could fill in ```temp``` then stop.

**Comment by mcornella** [comment]
Sorry, I misread (twice) your test suite. Here's a list of [completion options](http://zsh.sourceforge.net/Doc/Release/Options.html#Completion-4) for zsh, maybe you can find what you're looking for there while I'm trying to find the right combination for you.

**Comment by mcornella** [comment]
I've been messing around with settings and I haven't found a working setup. Perhaps you'll find better luck asking on the zsh mailing list... Sorry :sweat_smile:

**Comment by apjanke** [comment]
I believe the `zstyle ':completion:*' matcher-list` setting is what you're looking for. It's set [here in lib/completion.zsh](https://github.com/robbyrussell/oh-my-zsh/blob/192de6bcffb0294e19f4203f6f7dc1a7f3e427be/lib/completion.zsh#L13-L24) to do case-insensitive matching. You could override that by re-setting it in your `~/.zshrc` file after OMZ is loaded, changing it to a customized one which prefers exact-case matches to case-insensitive matches. (After, not before, or OMZ will just clobber it.)

I think this is what you're looking for, where same-case matches are preferred to differing-case matches. And [here's the doco](http://zsh.sourceforge.net/Doc/Release/Completion-Widgets.html#Completion-Matching-Control) if you want to customize it further. (Which you'll need to do if you want to get `Te` to fully match `Temp1` first, but offer `temp2` upon additional tab presses.)

```
zstyle ':completion:*' matcher-list 'r:|[._-]=* r:|=*' 'l:|=* r:|=*' 'm:{a-zA-Z}={A-Za-z}'
```

If you want to ignore case-differing matches entirely, you can set `CASE_SENSITIVE=true` in your `~/.zshrc` before loading OMZ.

**Comment by apjanke** [comment]
Similar to #4118, which may also be helpful.

**Comment by Joshfindit** [comment]
Alright; I did two things:

* Installed zsh by itself in a linux VM
* went to the mailing list

I found [this thread from 2003](http://www.zsh.org/mla/users/2003/msg00611.html) where Jesper Holmberg was looking for the same thing.

After going through all the suggestions here, and everything in that thread, the best I've come up with is this:

```
_cscomplete() {
  _complete
  return 1
}
zstyle ':completion:*:cscomplete:*' group-name case-sensitive
zstyle ':completion:*:complete:*' group-name case-insensitive
zstyle ':completion:*' group-order case-sensitive case-insensitive
zstyle ':completion:*::::' completer _cscomplete _complete
zstyle ':completion:*:cscomplete:*:*' matcher-list ''
zstyle ':completion:*:complete:*' matcher-list 'm:{a-z}={A-Z}'
```

Going back to the test case:

```
./Temp1
./temp2
./temp3
```
Typing ```cd te<tab>``` completes ```temp``` then stops. Pressing ```<tab>``` again does something interesting: 

```
temp2/      temp3/
Temp1/  temp2/  temp3/
```
It's a little messy, but does meet what I'm trying to do: suggest the case-sensitive completion first, and still offer the case-insensitive options.


*As with most things at this level, there is a gotcha:* if ```tEMP4``` exists, when ```te<tab>``` is typed, ```tEMP4``` will never be suggested, but that seems like a fair trade off.

I'd certainly welcome anyone who can refine this (and have reached out to the mailing list asking the same question), but in the mean time I hope this thread benefits the next person who asks this question.
